File: class10/python/TCC1001-ArvoreDerivacao.py
#!/usr/bin/env python3
"""TCC1001-ArvoreDerivacao.py | Aula 10
Gera árvore de derivação de uma GLC e detecta ambiguidade.
Gramática exemplo: E → E+E | E*E | (E) | id
"""
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class No:
    simbolo: str
    filhos: List["No"] = field(default_factory=list)

    def eh_folha(self) -> bool:
        return not self.filhos


def imprime_arvore(no: No, prefixo: str = "", eh_ultimo: bool = True) -> None:
    """Imprime a árvore em formato ASCII (estilo tree)."""
    connector = "└── " if eh_ultimo else "├── "
    print(prefixo + connector + no.simbolo)
    novo_prefixo = prefixo + ("    " if eh_ultimo else "│   ")
    for i, filho in enumerate(no.filhos):
        imprime_arvore(filho, novo_prefixo, i == len(no.filhos) - 1)


def mermaid(no: No) -> str:
    """Gera representação Mermaid (flowchart) da árvore."""
    linhas = ["flowchart TD"]
    contador = [0]

    def _id():
        contador[0] += 1
        return f"n{contador[0]}"

    def _walk(no: No, pid: Optional[str]):
        nid = _id()
        rotulo = no.simbolo if no.simbolo else "ε"
        linhas.append(f'    {nid}["{rotulo}"]')
        if pid:
            linhas.append(f"    {pid} --> {nid}")
        for f in no.filhos:
            _walk(f, nid)

    _walk(no, None)
    return "\n".join(linhas)


def arvore_id_mais_id() -> No:
    """Derivação leftmost de id+id: E ⇒ E+E ⇒ id+E ⇒ id+id"""
    return No("E", [
        No("E", [No("id")]),
        No("+"),
        No("E", [No("id")]),
    ])


def arvore_id_vezes_id_mais_id_esq() -> No:
    """id*(id+id) — agrupamento à esquerda: (id*id)+id"""
    return No("E", [
        No("E", [
            No("E", [No("id")]),
            No("*"),
            No("E", [No("id")]),
        ]),
        No("+"),
        No("E", [No("id")]),
    ])


def arvore_id_vezes_id_mais_id_dir() -> No:
    """id*id+id — agrupamento à direita: id*(id+id) — AMBIGUIDADE"""
    return No("E", [
        No("E", [No("id")]),
        No("*"),
        No("E", [
            No("E", [No("id")]),
            No("+"),
            No("E", [No("id")]),
        ]),
    ])


if __name__ == "__main__":
    print("=== Árvore 1: id+id (leftmost) ===")
    imprime_arvore(arvore_id_mais_id())

    print("\n=== Árvore 2: id*id+id — agrupamento à esquerda (id*id)+id ===")
    imprime_arvore(arvore_id_vezes_id_mais_id_esq())

    print("\n=== Árvore 3: id*id+id — agrupamento à direita id*(id+id) ===")
    imprime_arvore(arvore_id_vezes_id_mais_id_dir())

    print("\n>>> AMBIGUIDADE: as árvores 2 e 3 derivam a MESMA palavra 'id*id+id'")
    print(">>> com estruturas diferentes → gramática ambígua!\n")

    print("=== Mermaid da Árvore 1 ===")
    print(mermaid(arvore_id_mais_id()))

File: class10/java/TCC1001-ArvoreDerivacao.java
// TCC1001-ArvoreDerivacao.java | Aula 10
// Gera árvore de derivação de uma GLC e detecta ambiguidade.
// Gramática exemplo: E → E+E | E*E | (E) | id
// Espelho Java do TCC1001-ArvoreDerivacao.py (mesmas saídas).
import java.util.ArrayList;
import java.util.List;

public class TCC1001ArvoreDerivacao {

    static class No {
        String simbolo;
        List<No> filhos = new ArrayList<>();

        No(String simbolo) {
            this.simbolo = simbolo;
        }

        No add(String simbolo) {
            No f = new No(simbolo);
            filhos.add(f);
            return f;
        }

        boolean ehFolha() {
            return filhos.isEmpty();
        }
    }

    static int contador = 0;

    static void imprimeArvore(No no, String prefixo, boolean ehUltimo) {
        String connector = ehUltimo ? "└── " : "├── ";
        System.out.println(prefixo + connector + no.simbolo);
        String novoPrefixo = prefixo + (ehUltimo ? "    " : "│   ");
        for (int i = 0; i < no.filhos.size(); i++) {
            imprimeArvore(no.filhos.get(i), novoPrefixo, i == no.filhos.size() - 1);
        }
    }

    static void imprimeArvore(No no) {
        imprimeArvore(no, "", true);
    }

    static String mermaid(No raiz) {
        contador = 0;
        StringBuilder sb = new StringBuilder("flowchart TD\n");
        mermaidWalk(raiz, null, sb);
        return sb.toString();
    }

    static String mermaidWalk(No no, String pid, StringBuilder sb) {
        String nid = "n" + (++contador);
        String rotulo = no.simbolo.isEmpty() ? "ε" : no.simbolo;
        sb.append("    ").append(nid).append("[\"").append(rotulo).append("\"]\n");
        if (pid != null) {
            sb.append("    ").append(pid).append(" --> ").append(nid).append("\n");
        }
        for (No f : no.filhos) {
            mermaidWalk(f, nid, sb);
        }
        return nid;
    }

    static No arvoreIdMaisId() {
        // E ⇒ E+E ⇒ id+E ⇒ id+id
        No raiz = new No("E");
        No e1 = raiz.add("E"); e1.add("id");
        raiz.add("+");
        No e2 = raiz.add("E"); e2.add("id");
        return raiz;
    }

    static No arvoreIdVezesIdMaisIdEsq() {
        // (id*id)+id
        No raiz = new No("E");
        No eEsq = raiz.add("E");
        No eEsqE = eEsq.add("E"); eEsqE.add("id");
        eEsq.add("*");
        No eEsqD = eEsq.add("E"); eEsqD.add("id");
        raiz.add("+");
        No eDir = raiz.add("E"); eDir.add("id");
        return raiz;
    }

    static No arvoreIdVezesIdMaisIdDir() {
        // id*(id+id) — AMBIGUIDADE
        No raiz = new No("E");
        No eEsq = raiz.add("E"); eEsq.add("id");
        raiz.add("*");
        No eDir = raiz.add("E");
        No eDirE = eDir.add("E"); eDirE.add("id");
        eDir.add("+");
        No eDirD = eDir.add("E"); eDirD.add("id");
        return raiz;
    }

    public static void main(String[] args) {
        System.out.println("=== Árvore 1: id+id (leftmost) ===");
        imprimeArvore(arvoreIdMaisId());

        System.out.println("\n=== Árvore 2: id*id+id — agrupamento à esquerda (id*id)+id ===");
        imprimeArvore(arvoreIdVezesIdMaisIdEsq());

        System.out.println("\n=== Árvore 3: id*id+id — agrupamento à direita id*(id+id) ===");
        imprimeArvore(arvoreIdVezesIdMaisIdDir());

        System.out.println("\n>>> AMBIGUIDADE: as árvores 2 e 3 derivam a MESMA palavra 'id*id+id'");
        System.out.println(">>> com estruturas diferentes → gramática ambígua!\n");

        System.out.println("=== Mermaid da Árvore 1 ===");
        System.out.println(mermaid(arvoreIdMaisId()));
    }
}
